// protos/cache.proto
syntax = "proto3";

package cache;

// --- Original Cache Service (for clients) ---
service CacheService {
  rpc Get (GetRequest) returns (GetResponse) {}
  rpc Put (PutRequest) returns (PutResponse) {}
  rpc Delete (DeleteRequest) returns (DeleteResponse) {}
}

// --- New Replication Service (for primary to call replicas) ---
service ReplicationService {
  // Applies a replicated operation (Put or Delete)
  rpc ApplyOperation (ReplicationRequest) returns (ReplicationResponse) {}
}

// --- Messages for CacheService ---
message GetRequest { string key = 1; }
message GetResponse { string value = 1; bool found = 2; }
message PutRequest { string key = 1; string value = 2; }
message PutResponse { bool success = 1; }
message DeleteRequest { string key = 1; }
message DeleteResponse { bool success = 1; }

// --- Messages for ReplicationService ---
message ReplicationRequest {
  enum OperationType {
    PUT = 0;
    DEL = 1;
  }
  OperationType op_type = 1;
  string key = 2;
  string value = 3; // Only used for PUT operations
}

message ReplicationResponse {
  bool success = 1; // Did the replica apply it successfully?
}